src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/finance/TransactionSynchronisationService.kt

```diff
@@ -23,17 +26,44 @@ class TransactionSynchronisationService(
   private val financeService: FinanceApiService,
   private val telemetryClient: TelemetryClient,
   private val queueService: SynchronisationQueueService,
+  private val transactionIdBufferRepository: TransactionIdBufferRepository,
+  @Value("\${finance.transactions.forwardingDelaySeconds}")
+  private val forwardingDelaySeconds: Int,
 ) {
   private companion object {
     val log: Logger = LoggerFactory.getLogger(this::class.java)
   }
 
-  suspend fun transactionInserted(event: TransactionEvent, requestId: UUID) {
+  suspend fun transactionInsertCheck(event: TransactionEvent, requestId: UUID) {
+    if (event.isSourcedFromDPS()) {
+      telemetryClient.trackEvent("transactions-synchronisation-created-skipped", event.toTelemetryProperties())
+      return
+    }
+
+    if (transactionIdBufferRepository.existsById(event.transactionId)) {
+      log.debug("Skipping synchronisation of transaction ${event.transactionId} as id already recorded")
+      return
+    }
+
+    if (transactionIdBufferRepository.addId(event.transactionId) == 0) {
+      log.debug("Skipping synchronisation of transaction ${event.transactionId} as failed to record id")
+      return
+    }
+
+    queueService.sendMessage(
+      messageType = SynchronisationMessageType.PERFORM_TRANSACTION_SYNC.name,
+      synchronisationType = SynchronisationType.FINANCE,
+      message = EncapsulatedTransaction(event, requestId),
+      telemetryAttributes = mapOf("transactionId" to event.transactionId.toString()),
+      delaySeconds = forwardingDelaySeconds,
+    )
+  }
+
+  suspend fun transactionInserted(encapsulatedTransaction: InternalMessage<EncapsulatedTransaction>) {
```

Comment:
So when you have the hundreds of transactions scenario this will also be called hundred times all at the same time - but 5 seconds after the hundreds of xtag events are received?

Reply:
No, this is only called after the initial event successfully inserted the transaction id into the db table, and then queued the 5s delay message